fix: `--stdin-file-path` with nested configuration

cormacrelf opened the pull request:
<!--
	Thanks for submitting a Pull Request! We appreciate you spending the time to work on these changes.
	Please provide enough information so that others can review your PR.
	Once created, your PR will be automatically labeled according to changed files.
	Learn more about contributing: https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md
-->

## Summary

Fixes #8233. We were just missing a working_dir.join(), so biome was not able to find the project the stdin path was associated with. I don't know if that's exactly correct -- do we have to resolve the path of any symlinks or is cwd.join just fine?

<!-- If you create a user-facing change, please write a changeset: https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md#writing-a-changeset (your changeset is often a good starting point for this summary as well) -->

## Test Plan

I added an e2e test, and also improved the e2e test runner script. We can now:

- run without `cd e2e-tests` first
- filter tests: `./e2e-tests/test-all.sh stdin`, with glob matching
- Find out which test failed in red at the end instead of scrolling up through cargo output
- Use `{ "root": true }` in test configs because e2e-tests is now ignored from the root biome.json 

## Docs

This actually just makes the CLI behave as is already documented in the `stdin_file_path` CLI params.

cormacrelf committed:
Warn when --stdin-file-path is an ignored path

cormacrelf committed:
Add e2e-tests/stdin-nested-config

This tests the fix for the stdin with nested config issue.

cormacrelf committed:
Add a changeset

ematipico commented:
> Stdin has no relation to traversal mode or scanning

This, basically, changed the semantics of stdin. That's a breaking change. 

There have been discussions around it. However we can't take this change lightly because there are users out there that rely on it.

cormacrelf committed:
Fix stdin path not being joined with workdir

This fixes nested config discovery/matching with the filepath.

cormacrelf committed:
Ignore e2e-tests from root biome.json

This enables us to have explicit root: true without making the root biome project error out.

ematipico commented on crates/biome_cli/src/execute/std_in.rs:
```diff
@@ -29,6 +29,7 @@ pub(crate) fn run<'a>(
     let workspace = &*session.app.workspace;
     let console = &mut *session.app.console;
     let mut version = 0;
+    tracing::info!(?biome_path, ?project_key, "Running on stdin");
```
Not needed (project key is an internal information)

ematipico commented on crates/biome_cli/src/logging.rs:
```diff
@@ -33,7 +33,7 @@ pub fn setup_cli_subscriber(
         .transpose()
         .expect("Failed to create log file")
         .optional()
-        .or_else(std::io::stdout);
```
This is intentional. 
- stdout -> for machines
- stderr -> for humans

Since logs are structured and predictable, they are for machines, hence stdout 

Can you please revert? Even if we're wrong, it isn't part of the PR

cormacrelf replied:
I have better things to do than make and land another half dozen PRs so I guess you can submit separately if you care for it. But be aware your default log format is pretty, which is not machine readable, and any kind of stdout logging is incompatible with `--stdin-file-path`. It feels pretty silly to add `--log-file /dev/stderr`, something I have never had to type in my life for any other program.

ematipico commented on crates/biome_cli/src/logging.rs:
```diff
@@ -44,14 +44,15 @@ pub fn setup_cli_subscriber(
+        LoggingKind::Json => registry
+            .with(layer.json().flatten_event(true).with_filter(filter))
+            .init(),
+    };
```
What are we fixing here? a regression? would be nice to add it to the PR description and add a changeset for it

cormacrelf replied:
I couldn't fix this without the logs working so I fixed the logs. I documented this in the commit message, it is perhaps better to review my code commit by commit.

ematipico reviewed (changes requested):
Thank you! The bug fix looks solid, however I noticed some changes that aren't covered by the PR that should be reverted or explained in the PR description

cormacrelf committed:
Improve e2e test runner, can now run from workspace root and filter tests

This also runs the tests with bash, not sh. So we can do more advanced shell scripts
in the various test.sh.

cormacrelf commented on e2e-tests/test-all.sh:
```diff
@@ -1,12 +1,26 @@
+        if [[ "$x" != $FILTER ]]; then
+            echo "Skipping $x"
+            continue
+        fi
```
No it should not

ematipico reviewed (commented):
Overall, it looks good. I left a comment we should resolve, if it's applicable

cormacrelf commented:
rebased. this is done. CI markdown lint failure is pre-existing.

cormacrelf commented:
Would be nice to get this merged. It's ready to go.

ematipico committed:
Merge branch 'main' into fix-stdin-nested-config
